### ISSUE TITLE
One should not overwrite default make flags

### ISSUE THREAD

**Issue by bverhagen** [aphenriques/integral#9]
In (lib/integral/Makefile), CXXFLAGS and LDFLAGS get overwritten. However, a lot of tools and wrappers depend on make's default behaviour of respecting these environment variables in order to adapt the behaviour of make.

One use case of this is when the include and/or lib folder of your Lua installation is not in a system path.

A simple fix is to append to these variables instead of overwriting them.

**Comment by aphenriques** [comment]
Good call! Thanks!

**Comment by aphenriques** [comment]
@bverhagen, would you provide an example of a build tool that exports CXXFLAGS to the environment?

The solution in 75b7d6d641b74a973deb4ddc16b6b0b0f4df922e might be a little misleading in the following situation:

1. Given the Makefile:
```makefile
CXXFLAGS+=end!
all:
	@echo $(CXXFLAGS)
```
2. Invoking
```
make CXXFLAGS=begin
```

3. One would guess that it would produce `begin end!`, however the makefile will not append `end!` to `CXXFLAGS` and the actual output is:
```
begin
```

**Comment by bverhagen** [comment]
@aphenriques: Most package builders use it to set certain specific settings, like dpkg (all Debian-based Linux distro's) and makepkg (Arch Linux).

In addition, I use it to build Integral as a 3rd party dependency in Cmake using a non-system LuaJIT installation (also built by CMake as a 3rd party dependency):

`ExternalProject_Add(${PACKAGE}
        PREFIX ${MODULE}
        GIT_REPOSITORY https://github.com/aphenriques/integral.git
        GIT_SUBMODULES "dependencies/exception"
        CONFIGURE_COMMAND ""
        BUILD_COMMAND "make;CXX=${CMAKE_CXX_COMPILER};USE_LUAJIT=YES;LUAJIT_INCLUDE_FOLDER=${luajit_INCLUDE_DIRS};static"
        BUILD_IN_SOURCE 1
        INSTALL_COMMAND "make;--directory;${SOURCE_DIR};INSTALL_TOP=${INSTALL_DIR};USE_LUAJIT=YES;LUAJIT_INCLUDE_FOLDER=${luajit_INCLUDE_DIRS};CXX=${CMAKE_CXX_COMPILER};install_static;install_exception"
        UPDATE_COMMAND ""
        )`

Using the '+=' operator, make should append to the variable (as you describe with 'expected'). When testing my patch, I am pretty sure I saw the behaviour you describe as 'expected' and not the behaviour you describe as 'actual'.

I will look into this when I have the time.

**Comment by bverhagen** [comment]
@aphenriques : I did some experimentation.

According to the make manual, this:
```
make CXXFLAGS=begin
```
Will override all assigments of CXXFLAGS, unless the [override](https://www.gnu.org/software/make/manual/make.html#Override-Directive) keyword is used. I am not sure whether or not this keyword is GNU make specific though.

This, however:
```
CXXFLAGS=begin make
```
causes make to append to an environment variable, which yields the expected behaviour.

To elaborate on my previous answer: the above mentioned package builders indeed set specific environment variables to override this behaviour (as they have no idea which build tool is used) rather than make arguments.

Conclusion: I think the patch is valid as submitted and merged. I do have to change my CMake invocation to reflect the above behaviour by making my overrides environment variables rather than command line arguments.

**Comment by aphenriques** [comment]
@bverhagen Good point! Your patch is definitely the best way to go. I am still thinking about a better way to organize the makefiles flags. I will leave this issue open until I find the best solution.

By the way, if you are building integral as a static library, you do not need to pass any LDFLAGS. They are only used to build the shared library. So, I believe you could remove the LDFLAGS definition from your build command.

Thanks for your explanations!

**Comment by bverhagen** [comment]
You are right: I ended op not using the overwrite feature for the moment. I re-added the LDFLAGS as an additional, quick illustration to your question.

You can also consider moving to CMake, which is getting widely adopted in open source C++ libraries. It comes with its own quirks and peculiarities though. It has built-in support for conventions like appending CXXFLAGS and LDFLAGS from the environment.

**Comment by aphenriques** [comment]
I made some changes in the makefiles to preserve default make flags. They may break your build configurations.

Take a look at the Build section of README.md.

I changed the name of some make variables. In particular:

- USE_LUAJIT -> WITH_LUAJIT
- LUAJIT_INCLUDE_FOLDER -> LUA_INCLUDE_DIR

You may also want to build the library with OPTIMIZED=y and WITH_FPIC=n for increased performance.

I will stick with makefiles for now, if the build system gets messier, I will check CMake out.

Let me know what you think.

**Comment by bverhagen** [comment]
Hi @aphenriques,

Your new variables may indeed make more sense. Thank you for the tips to further improve it!

As for feedback on the Build section of the README: it is up to you, but I would not suggest users to actually overwrite buildflags: I would suggest to append to them (using the `CXXFLAGS='-std=c++17 -I/path/to/luajit/include' LDFLAGS=-L/path/to/luajit/lib LDLIBS=-lluajitlibrary make` syntax).

Users that override rather than append compiler options, have decided they know better what the library needs than its author(s), and are therefore on their own. These people already know they can do it this way in make anyway. It is not ideal w.r.t. maintenance either: changes in the compiler options in later versions will not be picked up by people overriding the options today.
This is just my 2 cents, it is up to you decide whether you concur to this.

Other than that, great additions!

**Comment by aphenriques** [comment]
@bverhagen , you are right, I changed the build section of the README (734a7552693bd46ea14e9b259f362199f5070616).

Experimenting with the new makefiles, it occurred to me that reading CXXFLAGS and LDFLAGS from the shell environment may pose a risk.

For example, I leave a tmux session in a remote VPS to work in a project that requires integral. I rarely open a new one. If some process changes those variables and I am not aware or I forget, I could get a different build than I am expecting. So I removed this feature.

I checked some projects and the oficial Lua source makefile and they do not look these variables in the environment. So I decided to leave these variables insulated from the shell in integral's makefile again.

There are pros and cons in this approach but it fits my needs better. Besides, integral built system is actually very simple. One could just grab all its sources and the `dependecies/exception/include/exception` directory, include Lua headers and library, and build. There are no OS or compiler dependent code in integral managed by the makefiles.
